web/src/lib/components/pages.tsx

```diff
@@ -237,47 +237,29 @@ export default class Pages extends Component<PagesProps, PagesState> {
     });
   }
 
-  private uploadRecordings(recordings: any[],
+  private async uploadRecordings(recordings: any[],
                            sentences: string[],
                            progressCb: Function): Promise<void> {
-
-    return new Promise<void>((res, rej) => {
-      this.ensurePrivacyAgreement().then(() => {
-        let runningTotal = 0;
-        let originalTotal = recordings.length;
-
-        // This function calls itself recursively until
-        // all recordings are uploaded.
-        let uploadNext = () => {
-          if (recordings.length === 0) {
-            this.props.api.uploadDemographicInfo().then(() => {
-              return;
-            });
-            res();
-            return;
-          }
-
-          let recording = recordings.pop();
-          let blob = recording.blob;
-          let sentence = sentences.pop();
-
-          this.props.api.uploadAudio(blob, sentence).then(() => {
-            runningTotal++;
-            let percentage = Math.floor((runningTotal / originalTotal) * 100);
-            progressCb && progressCb(percentage);
-            this.props.user.tallyRecording();
-            uploadNext();
-          });
-        };
-
-        // Start the recursive chain to upload the recordings serially.
-        uploadNext();
-      }).catch(rej);
-    }).then(() => {
+    try {
+      await this.ensurePrivacyAgreement();
+      const originalTotal = recordings.length;
+
+      for (let runningTotal = 1; runningTotal <= originalTotal; runningTotal++) {
+        const recording = recordings.pop();
+        const blob = recording.blob;
+        const sentence = sentences.pop();
+
+        await this.props.api.uploadAudio(blob, sentence);
+        let percentage = Math.floor((runningTotal / originalTotal) * 100);
+        progressCb && progressCb(percentage);
+        this.props.user.tallyRecording();
+      }
+      await this.props.api.uploadDemographicInfo();
+    } finally {
```

Comment:
We don't want this `finally`, and probably not even the `try ... catch` at all.

The code I had before is *really* bad. The way I used the Promise is hard to read, so I will just describe in words what the code was doing in the likely case it was unclear.

This function `uploadRecordings` ultimately determines weather or not we display an error message on the Record page. What happens is, the Record page calls this function (as an `onSubmit` callback prop), and if the function fails it assumed the problem came from `ensurePrivacyAgreement` (the old Promise structure made sure of this in a *very* hacky way) and pops up a dialog asking the user what to do with the clips.

The proper thing to do here would be for `ensurePrivacyAgreement` to throw a specific error type, and catch that in the Record page before displaying the dialog. But in the end I think it will be enough to simply remove the `try...catch` here and change the error message in the dialog.

Reply:
I think that should be changed in a separate PR though, so we don't mix logic changes with refactoring here

Reply:
@nevik we are already changing the logic in this refactor, that's what i was explaining. this change introduces a bug where if the user clicks on "don't accept" on the privacy policy, the robot turns into the wrong image on the background record page.

Reply:
Okay, I have more or less changed the part. Please verify if it works.